### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [2]:
# Dependencies and Setup
import pandas as pd

# File to Load (Remember to Change These)
file_to_load = "Resources/purchase_data.csv"

# Read Purchasing File and store into Pandas data frame
pd_df = pd.read_csv(file_to_load)
pd_df.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

## Player Count

* Display the total number of players


In [3]:
# creating variable for number of players to print out as a dataframe
total_players = len(pd.unique(pd_df["SN"]))
tp_df = pd.DataFrame({
    "Total Players":[total_players]})
tp_df


Total Players
0            576

## Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [4]:
# set variables of price and revenue by calculating the average and sums respectively
average_price = pd_df["Price"].mean()
total_revenue = pd_df["Price"].sum()

# set variables of purchaes and unique items by counting unique values of each 
total_purchases = len(pd.unique(pd_df["Purchase ID"]))
unique_items = len(pd.unique(pd_df["Item ID"]))

# created dataframe from values rendered above
pa_df = pd.DataFrame({"Number of Unique Items":[unique_items], 
                      "Average Price":[average_price],
                      "Number of Purchases":[total_purchases],
                      "Total Revenue": total_revenue})

# converted price and revenue to a currency value 
pa_df["Average Price"] = pa_df["Average Price"].map("${:,.2f}".format)
pa_df["Total Revenue"] = pa_df["Total Revenue"].map("${:,.2f}".format)

pa_df



Number of Unique Items Average Price  Number of Purchases Total Revenue
0                     179         $3.05                  780     $2,379.77

## Gender Demographics

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed




In [8]:
# created dataframe to show total counts and % of players by gender and put in ascending order of total count
gender_df = pd_df[["SN", "Gender", "Age"]].drop_duplicates()
gender_group = gender_df["Gender"].value_counts()

# calculated percent of gender
perc_gender = gender_group/total_players

# dataframe for gender breakdown by total count and % of players within each group 
renamed_gender = pd.DataFrame({"Total Count":gender_group, "Percentage of Players": perc_gender})

# changed number to percentage 
renamed_gender["Percentage of Players"] = renamed_gender["Percentage of Players"].map("{:,.2%}".format)
renamed_gender


Total Count Percentage of Players
Male                           484                84.03%
Female                          81                14.06%
Other / Non-Disclosed           11                 1.91%


## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender




* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [9]:
# Creating requested variables by grouping by gender 
gender_summary = pd_df.groupby("Gender").count()["SN"]
average_price = pd_df.groupby("Gender").mean()["Price"]
total_purchase = pd_df.groupby("Gender").sum()["Price"]
average_pp = total_purchase/gender_group

# Dataframe to reflect requested info 
renamed_summary = pd.DataFrame({"Purchase Count":gender_summary, "Average Purchase Price": average_price,
                                "Total Purchase Value":total_purchase, "Avg Total Purchase per Person":average_pp})

# changed number formatting to currency
renamed_summary["Average Purchase Price"] = renamed_summary["Average Purchase Price"].map("${:,.2f}".format)
renamed_summary["Total Purchase Value"] = renamed_summary["Total Purchase Value"].map("${:,.2f}".format)
renamed_summary["Avg Total Purchase per Person"] = renamed_summary["Avg Total Purchase per Person"].map("${:,.2f}".format)

renamed_summary



Purchase Count Average Purchase Price  \
Gender                                                         
Female                            113                  $3.20   
Male                              652                  $3.02   
Other / Non-Disclosed              15                  $3.35   

                      Total Purchase Value Avg Total Purchase per Person  
Gender                                                                    
Female                             $361.94                         $4.47  
Male                             $1,967.64                         $4.07  
Other / Non-Disclosed               $50.19                         $4.56

## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [10]:
# Looked at the oldest age in the array
print(pd_df["Age"].max())

# created buckets of age groups 
bins = [0, 9.99, 14.99, 19.99, 24.99, 29.99,34.99,39.99,50]
group_names = ["<10", "10-14", "15-19", "20-24", "25-29","30-34","35-39","40+"]
gender_df["Age Group"] = pd.cut(gender_df["Age"], bins, labels=group_names, include_lowest=True)

# variables for requested info in dataframe
age_group = gender_df["Age Group"].value_counts()
pp_players = age_group/total_players
gender_demo = pd.DataFrame({"Total Count":age_group,"Percentage of Players":pp_players})
gender_demo = gender_demo.sort_index()

# changed number formatting to percentage
gender_demo["Percentage of Players"] = gender_demo["Percentage of Players"].map("{:,.2%}".format)
gender_demo


45


Total Count Percentage of Players
<10             17                 2.95%
10-14           22                 3.82%
15-19          107                18.58%
20-24          258                44.79%
25-29           77                13.37%
30-34           52                 9.03%
35-39           31                 5.38%
40+             12                 2.08%

## Purchasing Analysis (Age)

* Bin the purchase_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [19]:
# created buckets of age groups 
bins = [0, 9.99, 14.99, 19.99, 24.99, 29.99,34.99,39.99,50]
group_names = ["<10", "10-14", "15-19", "20-24", "25-29","30-34","35-39","40+"]
pd_df["Age Group"] = pd.cut(pd_df["Age"], bins, labels=group_names, include_lowest=True)

# variables for dataframe
purchase_count = pd_df.groupby("Age Group").count()["Item ID"]
average_price = pd_df.groupby("Age Group").mean()["Price"]
total_purchase = purchase_count*average_price
average_purch_price = total_purchase/age_group
total_purchase = pd_df.groupby("Age Group").sum()["Price"]
average_pp = total_purchase/purchase_count
age_demo = pd.DataFrame({"Purchase Count":purchase_count, 
                         "Average Purchase Price": average_price,
                         "Total Purchase Value":total_purchase,
                         "Avg Total Purchase per Person":average_purch_price})

# changed number formatting to currency where appicable 
age_demo["Average Purchase Price"] = age_demo["Average Purchase Price"].map("${:,.2f}".format)
age_demo["Total Purchase Value"] = age_demo["Total Purchase Value"].map("${:,.2f}".format)
age_demo["Avg Total Purchase per Person"] = age_demo["Avg Total Purchase per Person"].map("${:,.2f}".format)
age_demo





Purchase Count Average Purchase Price Total Purchase Value  \
<10                23                  $3.35               $77.13   
10-14              28                  $2.96               $82.78   
15-19             136                  $3.04              $412.89   
20-24             365                  $3.05            $1,114.06   
25-29             101                  $2.90              $293.00   
30-34              73                  $2.93              $214.00   
35-39              41                  $3.60              $147.67   
40+                13                  $2.94               $38.24   

      Avg Total Purchase per Person  
<10                           $4.54  
10-14                         $3.76  
15-19                         $3.86  
20-24                         $4.32  
25-29                         $3.81  
30-34                         $4.12  
35-39                         $4.76  
40+                           $3.19

## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [28]:
# grouped by screename to get the necessary values 
purchase_count = pd_df.groupby("SN").count()["Item ID"]
average_purch_price = pd_df.groupby("SN").mean()["Price"]

# created variables and added sorting 
total_purchase = purchase_count*average_purch_price
top_spenders = pd.DataFrame({"Purchase Count":purchase_count,"Average Purchase Price":average_purch_price,"Total Purchase Value":total_purchase})
top_spenders_final = top_spenders.sort_values("Total Purchase Value",ascending=False)

# changed numerical vaues to currency
top_spenders_final["Average Purchase Price"] = top_spenders_final["Average Purchase Price"].map("${:,.2f}".format)
top_spenders_final["Total Purchase Value"] = top_spenders_final["Total Purchase Value"].map("${:,.2f}".format)

top_spenders_final.head(5)


Purchase Count Average Purchase Price Total Purchase Value
SN                                                                     
Lisosia93                 5                  $3.79               $18.96
Idastidru52               4                  $3.86               $15.45
Chamjask73                3                  $4.61               $13.83
Iral74                    4                  $3.40               $13.62
Iskadarya95               3                  $4.37               $13.10

## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, average item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [32]:
# grouped items by Item ID and Item Name by finding item's relative price and value by multiplying 
# by purchase count
purchase_count = pd_df.groupby(["Item ID","Item Name"]).count()["Price"]
total_value = pd_df.groupby(["Item ID", "Item Name"]).sum()["Price"]
price = total_value/purchase_count

# created dataframe to be ascending by purchase count
most_popular = pd.DataFrame({"Purchase Count":purchase_count,
                             "Item Price":price,
                             "Total Purchase Value":total_value})
most_popular_final = most_popular.sort_values("Purchase Count",ascending=False)

# changing number format to currency 
most_popular_final["Item Price"] = most_popular_final["Item Price"].map("${:,.2f}".format)
most_popular_final["Total Purchase Value"] = most_popular_final["Total Purchase Value"].map("${:,.2f}".format)
most_popular_final["Purchase Count"] = most_popular_final["Purchase Count"].astype(str).str.replace(".0","")
most_popular_final.head()


Purchase Count  \
Item ID Item Name                                                     
92      Final Critic                                             13   
178     Oathbreaker, Last Hope of the Breaking Storm             12   
145     Fiery Glass Crusader                                      9   
132     Persuasion                                                9   
108     Extraction, Quickblade Of Trembling Hands                 9   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
145     Fiery Glass Crusader                              $4.58   
132     Persuasion                                        $3.22   
108     Extraction, Quickblade Of Trembling Hands         $3.53   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
145     Fiery Glass Crusader                                       $41.22  
132     Persuasion                                                 $28.99  
108     Extraction, Quickblade Of Trembling Hands                  $31.77

## Most Profitable Items

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



In [18]:
# grouped by item ID and Item Name 
purchase_count = pd_df.groupby(["Item ID","Item Name"]).count()["Price"]

# created variables and put in a dataframe to ascending by Total Purchase Value
total_value = pd_df.groupby(["Item ID", "Item Name"]).sum()["Price"]
price = total_value/purchase_count

most_popular = pd.DataFrame({"Purchase Count":purchase_count,
                             "Item Price":price,
                             "Total Purchase Value":total_value})
most_popular_final = most_popular.sort_values("Total Purchase Value",ascending=False)

# changed number formatting to currency and removed tenths from Purchase Count column 
most_popular_final["Item Price"] = most_popular_final["Item Price"].map("${:,.2f}".format)
most_popular_final["Total Purchase Value"] = most_popular_final["Total Purchase Value"].map("${:,.2f}".format)
most_popular_final["Purchase Count"] = most_popular_final["Purchase Count"].astype(str).str.replace(".0","")
most_popular_final.head()

Purchase Count  \
Item ID Item Name                                                     
92      Final Critic                                             13   
178     Oathbreaker, Last Hope of the Breaking Storm             12   
82      Nirvana                                                   9   
145     Fiery Glass Crusader                                      9   
103     Singed Scalpel                                            8   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
82      Nirvana                                           $4.90   
145     Fiery Glass Crusader                              $4.58   
103     Singed Scalpel                                    $4.35   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
145     Fiery Glass Crusader                                       $41.22  
103     Singed Scalpel                                             $34.80

# Key Findings: 

- More than 4 out of 5 players are male
- Female and other genders spend more (on average per person) than male by about 10% and 12% respectively 
- Almost half (44%) of the players are between the ages of 20 and 24 and are the 3rd largest spending group among the 8 age group buckets 
